### ISSUE TITLE
[Feature Request] Adapt nexa-sdk CI scripts

### ISSUE THREAD

**Issue by iwr-redmond** [ddh0/easy-llama#9]
Congratulations on your rewrite nearing its first release!

Nexa SDK builds pre-compiled GPU wheels for most platforms in CI:

- Windows: CUDA, Vulkan
- Linux: CUDA, ROCm
- OSX: Metal

The scripts are located [here](https://github.com/NexaAI/nexa-sdk/tree/main/.github/workflows).

Please consider adapting these scripts for easy-llama, as installation remains a major challenge for ordinary users. On Windows, for example, one not only needs to have CUDA and Visual Studio installed but also copy some files from CUDA into MSVCC BuildCustomizations - an absolute mess!

**Comment by oobabooga** [comment]
See also:

* [oobabooga/llama-cpp-python-cuBLAS-wheels](https://github.com/oobabooga/llama-cpp-python-cuBLAS-wheels/tree/main/.github/workflows) (an up-to-date fork of [jllllll/llama-cpp-python-cuBLAS-wheels](https://github.com/jllllll/llama-cpp-python-cuBLAS-wheels/tree/main/.github/workflows))
* The [workflows](https://github.com/abetlen/llama-cpp-python/tree/main/.github/workflows) in llama-cpp-python, which also derive from jllllll's.

**Comment by ddh0** [comment]
Hey, thank both of you so much. For the next several days I'm mostly focused on the code itself, when it gets closer to v0.2.0 I'll definitely take a closer look at what you shared. I'll come back and update this thread when I start working on this.

**Comment by iwr-redmond** [comment]
Design-wise, you may, as an alternative, wish to consider the approach taken by the GPT4All Python package, which defaults to Vulkan and Metal (OSX), while providing CUDA as an add-on in the same package. As a result, there are no extra indexes required, although the Windows and Linux wheels are slightly larger (~120Mb each).

The GPT4All setup.py is [here](https://github.com/nomic-ai/gpt4all/blob/main/gpt4all-bindings/python/setup.py).

If you took this approach, you could either provide [rocm] and [cuda] extra options for the user to select, similar to GPT4All, _or_ use runtime code adapted from the [comfyui installable package](https://github.com/hiddenswitch/ComfyUI/blob/master/setup.py#L51) to auto-select the correct option.

Good luck finalizing the new version!

**Comment by ddh0** [comment]
Hi @iwr-redmond and @oobabooga, I hope you both don't mind the ping. :)

The code is coming along nicely - the core functionality i intended to replace from `llama-cpp-python` is implemented now:
- `ez.Llama`
  + `Llama.tokenize`
  + `Llama.token_to_piece`
  + `Llama.detokenize`
  + `Llama.generate_single`
  + `Llama.generate`
  + `Llama.stream`
  + `Llama.get_logits`
  + `Llama.get_scores`
  + `Llama.sample_greedy`
  + `Llama.sample`
- `ez.SamplerParams`
- `ez.SamplerPreset`
- `ez.SamplerPresets`
- `ez.PromptFormat`
- `ez.Thread`
  + `Thread.send`
  + `Thread.get_input_ids`
  + `Thread.as_string`
  + `Thread.add_message`
  + `Thread.warmup`
  + `Thread.interact`
- `ez.WebUI`
- `_LlamaStopwatch`

Just working on a few conveniences like built-in prompt formats.

Now I'm starting to turn my attention towards the publishing and distribution of the package. I see the CI [scripts](https://github.com/NexaAI/nexa-sdk/tree/main/.github/workflows) and llama-cpp-python [workflows](https://github.com/abetlen/llama-cpp-python/tree/main/.github/workflows), but to be honest I don't know what to do with those. I'm still pretty new to git, GitHub and PyPI, and I don't know what the best way forward would be.

Taking a step back, ideally I'd like to be able to have users just `pip install easy-llama` and everything would be taken care of. But I'm guessing that might not be realistic.

If either of you would be so kind as to point me in the right direction regarding the publishing and distribution of the package while handling different compute backends and operating systems, I'd be really thankful. In any case, I'll continue trying to figure this out.

**Comment by iwr-redmond** [comment]
> Taking a step back, ideally I'd like to be able to have users just pip install easy-llama and everything would be taken care of.

I would lean in the direction of adapting the [GPT4All setup.py](https://github.com/nomic-ai/gpt4all/blob/main/gpt4all-bindings/python/setup.py) linked above. You'll see on lines 9-61 that they build their shared libraries separately and then package them up into the Python library. GPT4All does work with `pip install gpt4all`, although for CUDA you use `pip install gpt4all[cuda]` because there is no detection code.

For inference, you could then extend the code [here](https://github.com/nomic-ai/gpt4all/blob/main/gpt4all-bindings/python/gpt4all/_pyllmodel.py#L67) to load the correct library (noting that Nomic use only Vulkan, Metal, and CUDA architectures).

**Comment by iwr-redmond** [comment]
Also, see [CIBuildWheel](https://github.com/pypa/cibuildwheel) which might be of assistance. It's designed specifically for building multi-platform wheels on GitHub for PyPI.

**Comment by ddh0** [comment]
Thanks very much, I'll check these out

**Comment by oobabooga** [comment]
Nice @ddh0, that's good to know. About GitHub workflows, it works like this:

1. You create the workflow files in the repository under `.github/workflows`
2. After that, when you click on the Actions tab in the repository, you will see entries on the left for each workflow:

![Image](https://github.com/user-attachments/assets/a5df585b-610a-44f9-a6ab-b7e8734c00d9)

3. Then, once you click on one of them, you will find a "Run workflow" button to start it:

![Image](https://github.com/user-attachments/assets/a9d39564-884b-4f52-b15b-6fb438532b3a)

Some comments on GitHub workflows from personal experience:

1. They can be unreliable: sometimes a workflow fails with an error, but if you run it again, it completes successfully.
2. They sometimes break as well. Examples: https://github.com/abetlen/llama-cpp-python/pull/1551 and https://github.com/abetlen/llama-cpp-python/pull/1894
3. GitHub has storage limits for wheels; uploading many variants (thousands) can trigger a payment method requirement at some point, though no actual charges were made for me personally.

I'm not familiar with the approaches mentioned by @iwr-redmond above, but it sounds nice to have all shared libraries at once. In a way, I think that's how PyTorch does things (for different GPU models rather than vendors), with the disadvantage that wheels become massive.

**Comment by ddh0** [comment]
Thanks so much for the detailed response, that's really helpful for me. I'll play with workflows and see if I can get something good working.

**Comment by iwr-redmond** [comment]
The Nomic folks have been updating their CI scripts for combined libraries & wheels, see [here](https://github.com/nomic-ai/gpt4all/blob/main/.circleci/continue_config.yml).

**Comment by iwr-redmond** [comment]
Closing as Llama is no longer compiled within this project.
